pkg/upgrade/upgrade.go

```diff
@@ -0,0 +1,230 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	apv1b2 "github.com/k0sproject/k0s/pkg/apis/autopilot/v1beta2"
+	k0sv1beta1 "github.com/k0sproject/k0s/pkg/apis/k0s/v1beta1"
+	"github.com/replicatedhq/embedded-cluster/kinds/apis/v1beta1"
+	clusterv1beta1 "github.com/replicatedhq/embedded-cluster/kinds/apis/v1beta1"
+	"github.com/replicatedhq/embedded-cluster/operator/pkg/autopilot"
+	"github.com/replicatedhq/embedded-cluster/operator/pkg/k8sutil"
+	"github.com/replicatedhq/embedded-cluster/operator/pkg/registry"
+	"github.com/replicatedhq/embedded-cluster/operator/pkg/release"
+	"github.com/replicatedhq/embedded-cluster/pkg/config"
+	"github.com/replicatedhq/embedded-cluster/pkg/helpers"
+	"github.com/replicatedhq/embedded-cluster/pkg/kotscli"
+	"github.com/replicatedhq/embedded-cluster/pkg/kubeutils"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	operatorChartName   = "embedded-cluster-operator"
+	upgradeJobName      = "embedded-cluster-upgrade-%s"
+	upgradeJobConfigMap = "upgrade-job-configmap-%s"
+)
+
+// Upgrade upgrades the embedded cluster to the version specified in the installation.
+// First the k0s cluster is upgraded, then addon charts are upgraded, and finally the installation is unlocked.
+func Upgrade(ctx context.Context, in *clusterv1beta1.Installation) error {
+	cli, err := k8sutil.KubeClient()
+	if err != nil {
+		return fmt.Errorf("failed to create kubernetes client: %w", err)
+	}
+
+	err = createInstallation(ctx, cli, in)
+	if err != nil {
+		return fmt.Errorf("apply installation: %w", err)
+	}
+
+	fmt.Printf("Upgrading to version %s\n", in.Spec.Config.Version)
+	err = k0sUpgrade(ctx, cli, in)
+	if err != nil {
+		return fmt.Errorf("k0s upgrade: %w", err)
+	}
+
+	// Augment the installation with data dirs that may not be present in the previous version.
+	// This is important to do ahead of updating the cluster config.
+	// We still cannot update the installation object as the CRDs are not updated yet.
```

Comment:
this is no longer true as we're using configmaps not CRDs. we should probably just remove `reApplyInstallation` and update the installation here directly as part of this or right after it